sqlglot/optimizer/scope.py

```diff
@@ -86,37 +86,93 @@ def _collect(self):
         self._derived_tables = []
         self._raw_columns = []
 
-        # We'll use this variable to pass state into the dfs generator.
-        # Whenever we set it to True, we exclude a subtree from traversal.
-        prune = False
-
-        for node, parent, _ in self.expression.dfs(prune=lambda *_: prune):
-            prune = False
-
+        for node, parent, _ in self.walk(bfs=False):
             if node is self.expression:
                 continue
-            if isinstance(node, exp.Column) and not isinstance(node.this, exp.Star):
+            elif isinstance(node, exp.Column) and not isinstance(node.this, exp.Star):
                 self._raw_columns.append(node)
             elif isinstance(node, exp.Table):
                 self._tables.append(node)
             elif isinstance(node, exp.UDTF):
                 self._derived_tables.append(node)
```

Comment:
We can merge this case with the one in line 100 by doing `isinstance(node, (exp.UDTF, exp.Subquery))`.